service/matching/tasklist/task_list_manager.go

```diff
@@ -407,67 +408,80 @@ func (c *taskListManagerImpl) RefreshTaskListPartitionConfig(ctx context.Context
+	if newConfig != nil {
+		// push update notification to all non-root partitions
+		c.notifyPartitionConfig(ctx, oldConfig, newConfig)
+	}
 	return nil
 }
 
-func (c *taskListManagerImpl) updatePartitionConfig(ctx context.Context, config *types.TaskListPartitionConfig) (int, types.TaskListPartitionConfig, error) {
+func (c *taskListManagerImpl) updatePartitionConfig(ctx context.Context, newConfig *types.TaskListPartitionConfig) (*types.TaskListPartitionConfig, *types.TaskListPartitionConfig, error) {
+	err := validatePartitionConfig(newConfig)
+	if err != nil {
+		return nil, nil, err
+	}
 	var version int64
-	numberOfPartitionsToRefresh := 1
 	c.partitionConfigLock.Lock()
 	defer c.partitionConfigLock.Unlock()
-	if c.partitionConfig != nil {
-		numberOfPartitionsToRefresh = int(c.partitionConfig.NumReadPartitions)
-		if isTaskListPartitionConfigEqual(*c.partitionConfig, *config) {
-			return 0, types.TaskListPartitionConfig{}, nil
+	oldConfig := c.partitionConfig
+	if oldConfig != nil {
+		if isTaskListPartitionConfigEqual(*oldConfig, *newConfig) {
+			return nil, nil, nil
 		}
-		version = c.partitionConfig.Version
+		version = oldConfig.Version
 	} else {
-		if config.NumReadPartitions == 1 && config.NumWritePartitions == 1 {
-			return 0, types.TaskListPartitionConfig{}, nil
+		if len(newConfig.ReadPartitions) == 1 && len(newConfig.WritePartitions) == 1 {
+			return nil, nil, nil
 		}
 	}
-	err := c.throttleRetry.Do(ctx, func() error {
-		return c.db.UpdateTaskListPartitionConfig(&persistence.TaskListPartitionConfig{
-			Version:            version + 1,
-			NumReadPartitions:  int(config.NumReadPartitions),
-			NumWritePartitions: int(config.NumWritePartitions),
-		})
+	err = c.throttleRetry.Do(ctx, func() error {
+		return c.db.UpdateTaskListPartitionConfig(toPersistenceConfig(version+1, newConfig))
 	})
 	if err != nil {
 		// We're not sure whether the update was persisted or not,
 		// Stop the tasklist manager and let it be reloaded
 		c.scope.IncCounter(metrics.TaskListPartitionUpdateFailedCounter)
 		c.Stop()
-		return 0, types.TaskListPartitionConfig{}, err
+		return nil, nil, err
 	}
 	c.partitionConfig = c.db.PartitionConfig().ToInternalType()
-	currentConfig := *c.partitionConfig
-	numberOfPartitionsToRefresh = common.MaxInt(numberOfPartitionsToRefresh, int(currentConfig.NumReadPartitions))
-	return numberOfPartitionsToRefresh, currentConfig, nil
+	return oldConfig, c.partitionConfig, nil
 }
 
-func (c *taskListManagerImpl) notifyPartitionConfig(ctx context.Context, config types.TaskListPartitionConfig, count int) {
+func (c *taskListManagerImpl) notifyPartitionConfig(ctx context.Context, oldConfig, newConfig *types.TaskListPartitionConfig) {
 	taskListType := types.TaskListTypeDecision.Ptr()
 	if c.taskListID.GetType() == persistence.TaskListTypeActivity {
 		taskListType = types.TaskListTypeActivity.Ptr()
 	}
+	// TODO: Do we want to notify partitions that were removed?
```

Comment:
we need to notify partitions that were removed from write

Reply:
Done.